### PR TITLE
Added Form

### PR DESCRIPTION
**Pull request by ericsoderberghp**
#### What does this PR do?

This adds a Form component and enhances the FormField component to make most forms easier to build in grommet.
This also simplifies the Button component to remove an interim Box when it isn't needed.

#### Where should the reviewer start?

form.stories.js for en example of how it is used

#### What testing has been done on this PR?

storybook

#### How should this be manually tested?

storybook

#### What are the relevant issues?

https://github.com/grommet/grommet/issues/2514

#### Do the grommet docs need to be updated?

yes

#### Should this PR be mentioned in the release notes?

yes

#### Is this change backwards compatible or is it a breaking change?

backwards compatible

### CODE REVIEW

**Comment by ericsoderberghp** [comment]
From slack conversations:

> Is it supposed to include form validation as well? Also please let it live inside grommet package - standalone packages tend to not be used as much as core grommet.

> i would also add my personal preference to have the validation rules as part of the FormField (so i have to look at a single field to see its type, name and validation) instead of looking at a large object of validation rules

> Also provide a HOC or react-hooks to easily ‘formify’ other input fields ie grommet-controls

> 1) How does the validation work?
2) Can we track the dirty flags of each field?

> 1. I really don’t like the `options` change into the FormField - imho choosing the input field should not be hardcoded part of FormField but rather in a HOC or react-hooks as i posted above.  Here is an example of integrating the Select using a SelectField component and withFormField https://github.com/atanasster/grommet-controls/blob/master/src/js/components/SelectField/SelectField.js
2. The validation rules should allow an array of validations, not just a function or regex- where the validations are chained one after the other. Example code for password validation https://github.com/atanasster/grommet-dashboard/blob/a09c3f543fd688bcb9c3386dba4bd28f70424353/components/profiles/SmallProfileForm.js#L46 and the running example https://grommet-dashboard.herokuapp.com/profile the smaller My Profile form, click submit and fulfill the validation rules one after the other
3. I found it makes it easier to have a collection of pre-defined validation rules, so user can start quickly: https://github.com/atanasster/grommet-controls/blob/master/src/js/components/validators/validators.js

> 4. A few nice to have for Form -
- onSubmitError/onInvalidForm event handlers to help display validation errors outside of the form itself.
- a facilitator option focusFirstChild (true/false) - if set auto focus the first field from the form so user can start typing right away
- a tag/as option (by default `form`) so user can switch to `div` - i had issues when nesting validations with warnings of not allowing nested forms.

> 1. name, options, hardcoding input fields do not belong to `FormField` - I agree 100%, the currently proposed `FormField` architecture is not correct - whats needed is a HOC (or similar) to wrap input fields instead of modifying `FormField`.
2. Do not create a grommet Form but instead use binding to formik/react-forms - in general a good idea, and thats the first thing I tried a year ago (I am a blackbelt in code-reuse), but at the time the available generic form libraries could not satisfy 100% my grommet needs and I ended up writing my own take on forms with grommet. It's possible things have changed in the meantime and something like formik might be a great lib to integrate in v2, but would need more accurate research.
3. Of course the easiest thing would be to refresh and re-use my own grommet forms code and add docs and examples - the design is pretty close to whats needed and I have been at it for close to a year now,  Also on my own its quite difficult to maintain a bunch of code and libraries, while I have my real projects still suffering from the v2 setbacks.

**Comment by ericsoderberghp** [comment]
I took an initial stab at comparing the approaches of this grommet Form component with using Formik ...

Here is an example with the grommet Form:
```
const FormExample = () => (
  <Form
    onSubmit={(value, { done }) => {
      // whatever submitting the form should entail
      done();
    }}
  >
    <FormField label="Name" name="name" required />
    <FormField label="Email" name="email" type="email" />
    <FormField label="Employee ID" name="employeeId" validate={/^[0-9]+$/} />
    <FormField
      label="Size"
      name="size"
      component={Select}
      options={['small', 'medium', 'large']}
    />
    <FormField label="Comments" name="comments" component={TextArea} />
    <Button type="submit" label="Update" />
  </Form>
);
```

And here is an equivalent example using Formik:
```
class FormikExample extends Component {
  render() {
    return (
      <Formik
        validate={values => {
          const errors = {};
          if (!values.name) {
            errors.name = 'required';
          }
          if (!values.employeeId) {
            errors.employeeId = 'required';
          } else if (!errors.employeeId.match(/^[0-9]+$/)) {
            errors.employeeId = 'invalid';
          }
        }}
        onSubmit={this.onSubmit}
      >
        {({
          values,
          errors,
          handleChange,
          handleSubmit,
        }) => (
          <form onSubmit={handleSubmit}>
            <FormField label="Name" error={errors.name}>
              <TextInput name="name" value={values.name} onChange={handleChange} />
            </FormField>
            <FormField label="Email" error={errors.email}>
              <TextInput name="email" type="email" value={values.email} onChange={handleChange} />
            </FormField>
            <FormField label="Employee ID" error={errors.employeeId}>
              <TextInput name="employeeId" value={values.employeeId} onChange={handleChange} />
            </FormField>
            <FormField label="Size" error={errors.size}>
              <Select
                name="size"
                options={['small', 'medium', 'large']}
                value={values.size}
                onChange={handleChange}
              />
            </FormField>
            <FormField label="Comments" error={errors.comments}>
              <TextArea name="comments" value={values.comments} onChange={handleChange} />
            </FormField>
            <Button type="submit" label="Update" />
          </form>
        )}
      </Formik>
    );
  }
}
```

A few observations:
- Formik can be used with grommet.
- Because of the grommet FormField + TextInput pattern, it doesn't work to leverage Field from Formik. One option to address this is to create separate components for each grommet input component that wraps it with FormField and is built to work with Formik's Field.
- The grommet Form example is 19 lines long. The Formik example is 50 lines long.

File: src/js/components/Form/Form.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      value !== priorValue ||
+      errors !== priorErrors ||
+      messages !== priorMessages
+    ) {
+      return {
+        value: value !== priorValue ? value : stateValue,
+        priorValue: value,
+        errors: (errors !== priorErrors ? errors : stateErrors) || {},
+        priorErrors: errors,
+        messages: deepMerge(defaultMessages, messages),
```
Comment (alansouzati) [inline src/js/components/Form/Form.js]: why do we need a deepMerge here ?

File: src/js/components/Form/Form.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  state = { errors: {}, value: {}, touched: {} };
+
+  validations = {};
+
+  onSubmit = event => {
+    const { onSubmit } = this.props;
+    const { errors, value } = this.state;
+    event.preventDefault();
```
Comment (alansouzati) [inline src/js/components/Form/Form.js]: can we add a comment explaining why event.preventDefault is here?

File: src/js/components/FormField/doc.js [hunk 3]
```diff
@@ -28,6 +28,23 @@ export const doc = FormField => {
...
+    ),
+    pad: PropTypes.bool.description(
+      'Whether to add padding to align with the padding of TextInput.',
+    ),
+    required: PropTypes.bool.description('Whether the field is required.'),
+    validate: PropTypes.oneOfType([
+      PropTypes.object, // regular expression
+      PropTypes.func,
+    ]).description(
+      `Validation rule. Provide a regular expression or a function. If a
```
Comment (alansouzati) [inline src/js/components/FormField/doc.js]: can we expand the propType to be a shape: 
```
PropTypes.shape({ regexp: RegExp, message: PropTypes.string })
```

I don't think React has a RegExp proptype. maybe we can add "any" and explain that regexp should be a regexp instance ?

File: src/js/components/Form/__tests__/Form-test.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      <Grommet>
+        <Form>
+          <FormField name="test" />
+        </Form>
+      </Grommet>,
+    );
+    const tree = component.toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+});
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Form/__tests__/Form-test.js]: done

File: src/js/components/Form/Form.js (hunk 1 above)
Comment (ericsoderberghp) [reply src/js/components/Form/Form.js]: I was allowing the caller to pass sparse `messages`.

File: src/js/components/Form/Form.js (hunk 2 above)
Comment (ericsoderberghp) [reply src/js/components/Form/Form.js]: done

File: src/js/components/Form/Form.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  };
+
+  render() {
+    const { children, ...rest } = this.props;
+    const { errors, touched, value, messages } = this.state;
+    return (
+      <form {...rest} onSubmit={this.onSubmit}>
+        <FormContext.Provider
+          value={{
+            addValidation: this.addValidation,
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Form/Form.js]: I think this is the general recommendation for how a context can be used in react.

File: src/js/components/Form/Form.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            errors,
+            messages,
+            touched,
+            update: this.update,
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Form/Form.js]: I think this is the general recommendation for how a context can be used in react.

File: src/js/components/Form/README.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+```
+{
+
+}
+```
+
+**messages**
+
+Custom messages. Used for accessibility by screen readers. Defaults to `{
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/Form/README.md]: Agreed

File: src/js/components/FormField/README.md [hunk 8]
```diff
@@ -47,4 +47,41 @@ A short label describing the field
...
+
+Whether the field is required.
+
+```
+boolean
+```
+
+**validate**
+
+Validation rule. Provide a regular expression or a function. If a
```
[the remark this answers is not in the document]
Comment (ericsoderberghp) [reply src/js/components/FormField/README.md]: done

File: src/js/components/FormField/doc.js (hunk 3 above)
Comment (ericsoderberghp) [reply src/js/components/FormField/doc.js]: done

File: src/js/components/Form/Form.js (hunk 1 above)
Comment (alansouzati) [reply src/js/components/Form/Form.js]: i hear you, but does it need to be "deep"?

Comment (ericsoderberghp) [reply src/js/components/Form/Form.js]: doh, of course not. Fixed.

File: src/js/components/FormField/README.md [hunk 9]
```diff
@@ -47,4 +47,45 @@ A short label describing the field
...
+      function is provided, it will be called with two arguments, the value
+      for this field and the entire value object. This permits validation to
+      encompass multiple fields. The function should return a string message
+      indicating describing the validation issue, if any.
```
Comment (alansouzati) [inline src/js/components/FormField/README.md]: remove indicating ?

Comment (ericsoderberghp) [reply src/js/components/FormField/README.md]: done

**Pull request merged**